descwl_coadd/online_coadds.py

```diff
@@ -0,0 +1,173 @@
+"""
+TODO
+    - actually try to run it
+    - psf coadds
+    - noise coadds; may want to generate the noise images on the fly
+      within this code
+"""
+import lsst.afw.math as afw_math
+import lsst.afw.image as afw_image
+from lsst.pipe.tasks.accumulatorMeanStack import AccumulatorMeanStack
+from lsst.pipe.tasks.assembleCoadd import AssembleCoaddTask
+from lsst.daf.butler import DeferredDatasetHandle
+
+INTERP = 'lanczos3'
+
+
+def make_online_coadd(exps, coadd_wcs, coadd_bbox):
+    """
+    make a coadd from the input exposures, working in "online mode",
+    adding each exposure separately.  This saves memory when
+    the exposures are being read from disk
+
+    Parameters
+    ----------
+    exps: list
+        Either a list of exposures or a list of DeferredDatasetHandle
+    coadd_wcs: DM wcs object
+        The target wcs
+    coadd_bbox: geom.Box2I
+        The bounding fox for the coadd
+
+    Returns
+    -------
+    ExposureF for coadd
+    """
+    coadd_exp = afw_image.ExposureF(coadd_bbox, coadd_wcs)
+
+    coadd_stats_ctrl = get_coadd_stats_control(
+        afw_image.Mask.getPlaneBitMask('EDGE')
+    )
+
+    stacker = make_online_stacker(coadd_stats_ctrl, coadd_bbox)
+
+    warp_config = afw_math.Warper.ConfigClass()
+    warp_config.warpingKernelName = INTERP
+    warper = afw_math.Warper.fromConfig(warp_config)
+
+    for texp in exps:
+        if isinstance(texp, DeferredDatasetHandle):
+            exp = texp.get()
+        else:
+            exp = texp
+
+        weight = get_exp_weight(exp)
+
+        wexp = warper.warpExposure(
+            coadd_wcs,
+            exp,
+            maxBBox=exp.getBBox(),
+            destBBox=coadd_bbox,
+        )
+
+        stacker.add_masked_image(wexp.image, weight=weight)
+
+        # if isinstance(texp, DeferredDatasetHandle):
+        #     # is this needed?
+        #     del exp
+
+    stacker.fill_stacked_masked_image(coadd_exp.image)
+
+    return coadd_exp
+
+
+def make_online_stacker(stats_ctrl, coadd_bbox):
+    """
+    make an AccumulatorMeanStack to online coadding
+
+    Parameters
+    ----------
+    stats_ctrl: afw_math.StatisticsControl
+        The stats control
+    coadd_bbox: geom.Box2I
+        The coadd bbox
+    """
+
+    # dims is C/numpy ordering
+    dims = get_dims_from_bbox(coadd_bbox)
+
+    mask_map = AssembleCoaddTask.setRejectedMaskMapping(stats_ctrl)
+
+    return AccumulatorMeanStack(
+        shape=dims,
```

Comment:
I'm not sure why you do the whole bbox dance; if you send the coadd down the pipe you can get the array dimensions from there without worrying about the row/column flipping.